### DISCUSSION TITLE
مشکل در ارسال درخواست به سمت درگاه زرین پال

### DISCUSSION THREAD

**Discussion by AliGod8001** [Sina-Soltani/Parbad#344] [Q&A]
من روی لوکال تمامی کد های مربوط به ارسال درخواست به سمت درگاه رو پیاده سازی کردم برای تست که درصورت اکی بودن ببرم روی محیط اصلی ولی کد ها با ارور مواجه میشه و پیش نمیره 
ارور مربوط به ارسال درخواست با _onlinePayment.RequestAsync و متن ارور هم "An unknown error is happened." هست

کد های اصلی
`
var callBackUrl = Url.Action("Callback", "Payment", null, Request.Scheme, Request.Host.Value);
var response = await _onlinePayment.RequestAsync(invoice =>
{
    invoice.SetAmount(transaction.Price * 10)
        .SetCallbackUrl(callBackUrl)
        .SetTrackingNumber(transaction.Id)
        .SetZarinPalData("شارژ کیف پول")
        .UseZarinPal();
});;

if (response.Status == PaymentRequestResultStatus.Succeed)
{
    await response.GatewayTransporter.TransportAsync();
}
`

متد کال بک
`
public async Task<IActionResult> Callback()
{
    var invoice =await _onlinePayment.FetchAsync();
    var transaction = await _transactionService.GetTransaction((int)invoice.TrackingNumber);
    if (transaction == null)
    {
        return NotFound();
    }
    
    if (invoice.IsSucceed)
    {
        var verifyResult=await _onlinePayment.VerifyAsync(invoice);
        if (verifyResult.IsSucceed)
        {
            var user = await _userService.GetUser(transaction.UserId);
            transaction.TransactionStatus = TransactionStatus.Success;
            transaction.TrackingNumber = verifyResult.TransactionCode;
            transaction.TransactionTime = DateTime.Now;
            transaction.WalletRemain = user.WalletBalance + transaction.Price;
            transaction.Gateway = Gateway.Zarinpal;
            transaction.ModificationDate = DateTime.Now;

            user.WalletBalance = transaction.WalletRemain;
            
            var rep = await _repositoryService.GetRepository();
            var repTransaction = new RepTransaction()
            {
                TransactionTitle = Domains.Repository.TransactionTitle.WalletRecharge,
                TransactionType = Domains.Repository.TransactionType.Deposit,
                price = transaction.Price,
                TransactionDateTime = DateTime.Now,
                RefCode = new Utils().GenerateRefCode(20),
                text = $"شارژ کیف پول به مبلغ {transaction.Price.ToToman()} هزارتومان توسط {transaction.User.FullName}",
                Repository = rep,
                Transaction = transaction
            };
            rep.Balance = rep.Balance + transaction.Price;
            rep.ModificationDate = DateTime.Now;
            await _repTransactionService.AddTransaction(repTransaction);
            transaction.RepTransaction = repTransaction;
        }
            
        var rows = await _context.SaveChangesAsync();
        if (rows > 0)
        {
            TempData["errorPaymentTrackingNumber"] = invoice.TrackingNumber;
            return RedirectToAction("Error");
        }

        return NotFound();
    }

    transaction.TransactionStatus = TransactionStatus.Failed;
    transaction.ModificationDate = DateTime.Now;
    var errorRows = await _context.SaveChangesAsync();
    if (errorRows > 0)
    {
        RedirectToAction("Success");
    }
    return NotFound();
}
`

کد های program.cs 
`
builder.Services.AddParbad()
    .ConfigureGateways(gateways =>
    {
        gateways.AddZarinPal()
            .WithAccounts(accounts =>
            {
                accounts.AddInMemory(account =>
                {
                    account.IsSandbox = bool.Parse(builder.Configuration["GateWays:ZarinPal:SandBox"]);
                    account.MerchantId = builder.Configuration["GateWays:ZarinPal:MerchantId"];
                    account.Name = "DefaultZarinPal";
                });
            });
    })
    .ConfigureStorage(storage =>
    {
        storage.UseMemoryCache();
    })
    .ConfigureHttpContext(b =>
    {
        b.UseDefaultAspNetCore();
    });
`

**Comment by Sina-Soltani** [comment]
درود
لطفا لاگ خطا رو ارسال کنید

**Reply by AliGod8001** [reply]
status ریسپانس رو failed میده 

`
                    var callBackUrl = Url.Action("Callback", "Payment", null, Request.Scheme, Request.Host.Value);
                    var response = await _onlinePayment.RequestAsync(invoice =>
                    {
                        invoice.SetAmount(transaction.Price * 10)
                            .SetCallbackUrl(callBackUrl)
                            .SetTrackingNumber(transaction.Id)
                            .SetZarinPalData("شارژ کیف پول")
                            .UseZarinPal();
                    });;
                    
                    if (response.Status == PaymentRequestResultStatus.Succeed)
                    {
                        await response.GatewayTransporter.TransportAsync();
                    }
`

response.IsSucceed = false;
response.Message = "An unknown error is happened."

**Reply by Sina-Soltani** [reply]
درود
کد رو قبلا اینجا قرار داده بودید نیازی نبود دوباره ارسال کنید.
لاگ خطا رو ارسال کنید لطفا.

**Reply by AliGod8001** [reply]
پیام ریسپانس رو که لاگ میگیرم این توی لاگ میاد که قبلا هم اشاره کردم بهش
"An unknown error is happened." 
این محتوای message ریسپانس بالا هست

**Reply by Sina-Soltani** [reply]
درود.
دوست عزیز، لاگ یعنی Logger
اگر نیاز به پراپرتی message برای عیب یابی داشته باشیم، بنده هم ذکر میکردم مقدار پراپرتی message رو بزارید.
برنامه شما احتمالا Logger نداره اصلا، شاید دلیل اینکه آشنا نیستید با لاگر همین باشه.
در هر صورت بدون لاگر کمکی از من بر نمیاد، چون مشخص نیست خطا چه چیزی هست.

**Reply by AliGod8001** [reply]
درود
متوجه منظورتون هستم
در حال حاضر از سرویس elmah برای لاگ توی سیستم داره استفاده میشه ولی مشکل اینجاست که هیچ لاگی نمیندازه و قبلا هم عرض کردم
هیچ لاگی نمی افته و وضعیت ناموفق میخوره که بسیار عجیبه این مورد

**Reply by Sina-Soltani** [reply]
الماه یک لاگر کاملا مجزا از لاگر اصلی دات نت هست و شما هیچگونه لاگی که توسط دات نت یا محصولاتی که از لاگر دات نت استفاده میکنند (مثل این پکیج)  رو مشاهده نخواهید کرد.
بنابراین شما باید الماه رو هماهنگ کنید با لاگر اصلی دات نت (فکر میکنم یک نوگت مخصوص برای این کار وجود داره. من آشنایی ندارم)
در غیر اینصورت تا زمانی که لاگ خطا مشاهده نشه، نمیشه گفت که خطا دقیقا چه چیزی هست. اون پیغام موجود در پراپرتی مسیج، صرفا برای کاربر یا نمایش در برنامه هست، نه برای استفاده به عنوان لاگ و خطایابی.

**Reply by AliGod8001** [reply]
درود
این متن اکسپشن هست که فکر کنم درگاهی با این اسم پیدا نکرده درحالی که من درگاه زرین پالم رو با همین اسم دقیقا توی پروگرم اضافه کردم
Parbad.Exceptions.GatewayNotFoundException: No gateway found with the name DefaultZarinPal. Make sure you have already added the required gateways. To add and configure a gateway use the ConfigureGateways method.
   at Parbad.Internal.DefaultGatewayProvider.Provide(String gatewayName)
   at Parbad.Internal.DefaultOnlinePayment.RequestAsync(Invoice invoice, CancellationToken cancellationToken)
   at Parbad.OnlinePaymentExtensions.RequestAsync(IOnlinePayment onlinePayment, Action`1 configureInvoice, CancellationToken cancellationToken)
   at Hadi.Controllers.PaymentController.Pay(Int32 price, String text) in F:\Projects\Hadi\Hadi\Hadi\Controllers\PaymentController.cs:line 67

**Reply by Sina-Soltani** [reply]
نام درگاه در قسمت تنظیمات چیزی نیست که توسط کاربر قابل تنظیم یا تغییر باشه. در واقع اصلا چنین قابلیتی در این سیستم حتی وجود هم نداره.
با توجه به کدهای مربوط به قسمت ایجاد درخواست، که در اینجا قرار دادید، این خطا به هیچ وجه نمیتونه اتفاق بیوفته. مگر اینکه کدهایی که در اینجا قرار دادید کدهایی نباشه که واقعا دارید استفاده می کنید.

**Reply by AliGod8001** [reply]
معذرت برای اینکه اکسپشن رو دقیق تر متوجه بشم از داک خودتون روش ارسال درخواست رو تغییر دادم و با ارور بالا مواجه شدم و حلش کردم الان به این اکسپشن برخوردم
کد بخش ارسال درخواست
`
     var callBackUrl = Url.Action("Callback", "Payment", null, Request.Scheme, Request.Host.Value);
                var result = await _onlinePayment.RequestAsync("Zarinpal", transaction.Id, transaction.Price*10, callBackUrl);
`

متن ارور 
Request failed. ZarinPal Gateway needs invoice information. Please use the SetZarinPalData method of Invoice Builder to add the information.

**Reply by Sina-Soltani** [reply]
متن خطا داره مشکل رو به وضوح به شما میگه.
برای استفاده از درگاه زرین پال نیاز به یک متن مانند توضیح خرید هست که باید برای زرین پال ارسال بشه هنگام ارسال درخواست.
برای این کار باید از متد SetZarinPalData استفاده کنید

**Reply by AliGod8001** [reply]
مشکل حل شد ظاهرا خطا برای استفاده از حالت سندباکس بود.
درگاه رو که از حالت سندباکس خارج میکنم اوکی میشه و کدها کار میکنه

فقط اون بخشی که کاربر رو به لینک درگاه هدایت میکنه ارور CORS میده و کاربر هدایت نمیشه
این مورد مربوط به شما میشه یا درگاه ؟
یا هیچ کدوم ؟

**Reply by Sina-Soltani** [reply]
خیر خطای CORS به درگاه یا این پکیج ارتباطی نداره.
باید تنظیمات CORS رو در برنامتون فعال کنید.

**Reply by AliGod8001** [reply]
تشکر ویژه از شما به خاطر همراهیتون 🙏🏻

**Reply by Sina-Soltani** [reply]
خواهش میکنم.
